File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.py
import heapq

def solution(n, costs):
    answer = 0
    
    edges = []
    for a, b, cost in costs:
        edges.append((cost, a, b))
    heapq.heapify(edges) # 비용 기준 간선 오름차순 정렬
    
    parent = [i for i in range(n)] # 유니온 파인드용 부모 배열
    
    def find(x):
        if parent[x] != x:
            parent[x] = find(parent[x])
        return parent[x]
    def union(a, b):
        root_a = find(a)
        root_b = find(b)
        
        if root_a != root_b:
            parent[root_b] = root_a
            return True
        return False # 둘이 이미 같은 집합 내에 있었음
    
    while edges:
        cost, a, b = heapq.heappop(edges)
        if union(a, b): # 둘이 합쳐짐 -> 사이클을 생성하지 않아 선택 가능
            answer += cost
    
    return answer

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.java
import java.util.*;

class Solution {
    static int[] parents;
    
    public int solution(int n, int[][] costs) {
        int answer = 0;
        
        // 간선 가중치 오름차순 정렬
        Arrays.sort(costs, (a, b) -> Integer.compare(a[2], b[2]));
        
        // union-find용
        parents = new int[n];
        for (int i = 0; i < n; i++) parents[i] = i;
        
        int edgeCount = 0; // 선택한 간선 수
        
        for (int[] edge: costs) {
            int a = edge[0], b = edge[1], cost = edge[2];
            
            if (union(a, b)) { // 둘이 합쳐진다면 선택 가능
                answer += cost;
                edgeCount++;
                
                if (edgeCount == n-1) break; // 선택한 간선의 수가 n-1 개가 되면 종료
            }
        }
        
        return answer;
    }
    
    static int find(int x) {
        if (parents[x] != x) {
            parents[x] = find(parents[x]);
        }
        return parents[x];
    }
    
    static boolean union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        
        if (rootX != rootY) {
            parents[rootY] = rootX;
            return true;
        }
        return false;
    }
}